### PR TITLE
rpc: Serialize in getblock without cs_main

### LINKED ISSUE #15925: getblock performance issue on verbosity
`getblock` is slow on verbosity, it's happened because `LOCK(cs_main)` applied for whole function including JSON generation
https://github.com/bitcoin/bitcoin/blob/ef70f9b52b851c7997a9f1a0834714e3eebc1fd8/src/rpc/blockchain.cpp#L810

I added few lines which print spent time:
```cpp
    LOCK(cs_main);
    auto t1 = std::chrono::high_resolution_clock::now();
...
    auto t2 = std::chrono::high_resolution_clock::now();
    auto ret = blockToJSON(block, pblockindex, verbosity >= 2);
    auto t3 = std::chrono::high_resolution_clock::now();
    std::cout << "blockToJSON " << pblockindex->nHeight << ": " << std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1).count() << " " << std::chrono::duration_cast<std::chrono::microseconds>(t3 - t1).count() << std::endl;
    return ret;
```

and numbers looks like:
```
blockToJSON 570019: 14727 201870
blockToJSON 570022: 16049 225224
blockToJSON 570021: 33776 217398
blockToJSON 570020: 21754 191669
blockToJSON 570023: 30949 236941
blockToJSON 570025: 23346 223673
blockToJSON 570024: 18732 203002
blockToJSON 570026: 17467 216213
blockToJSON 570027: 16658 217395
blockToJSON 570028: 19015 219183
```

i.e. everything before JSON generation takes 15-30ms, while JSON takes minimum 190ms and lock `cs_main` for all this time. As result it's possible extract only ~5 blocks per second.

Looking on the code I think it's possible copy data from `CBlockIndex` to extra structure, so in this case `cs_main` can be unlocked right after `GetBlockChecked`.
https://github.com/bitcoin/bitcoin/blob/ef70f9b52b851c7997a9f1a0834714e3eebc1fd8/src/rpc/blockchain.cpp#L828
As result `getblock` in verbosity mode should be much faster (in case if `-rpcthreads` will be adjusted too).

Does this make sense?

### PR DESCRIPTION
**Pull request by maflcko**
No need to hold cs_main when serializing a struct to json

Fixes: #15925

### CODE REVIEW

**Comment by jnewbery** [comment]
Does it make sense to add an `AssertLockNotHeld(cs_main)` to those functions?

**Comment by promag** [comment]
For reference https://github.com/bitcoin/bitcoin/pull/12153#issuecomment-452087924.

Concept ACK.

> Does it make sense to add an `AssertLockNotHeld(cs_main)` to those functions?

Maybe `EXCLUDES` annotation instead?

**Comment by laanwj** [comment]
Concept ACK. The whole point of moving the lock of `cs_main` inside the individual RPC functions was to allow optimizations such as this.

**Comment by MarcoFalke** [comment]
Added lock annotations as requested by @promag

**Comment by jnewbery** [comment]
utACK faea56400d5578023133cf4d1c761cdeb0c3e3da

**Comment by fanatid** [comment]
Why not use `ENTER_CRITICAL_SECTION` & `LEAVE_CRITICAL_SECTION` instead `LOCK`?

**Comment by promag** [comment]
Because a scoped lock is more clear and "secure".

**Comment by jonasschnelli** [comment]
utACK faea56400d5578023133cf4d1c761cdeb0c3e3da

File: src/rpc/blockchain.h [hunk 1]
```diff
@@ -5,9 +5,13 @@
...
 
-#include <vector>
-#include <stdint.h>
 #include <amount.h>
+#include <sync.h>
+
+#include <stdint.h>
+#include <vector>
+
+extern RecursiveMutex cs_main;
```
Comment (elichai) [inline src/rpc/blockchain.h]: @MarcoFalke Do you use `RecursiveMutex` instead of `CCriticalSection` on purpose?

**Pull request merged**